### PR TITLE
Secure the contact_form from spammers

### PR DESCRIPTION
**Pull request by hibatallahAouadni**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.6.1.x
| Description?  | I adopt the solution in this [link](http://www.nfriedly.com/techblog/2009/11/how-to-build-a-spam-free-contact-forms-without-captchas/) to solve the problem of spamming the contact form
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/PSCSX-9132
| How to test?  |

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8168)

### CODE REVIEW

File: controllers/front/ContactController.php [hunk 1]
```diff
@@ -36,9 +36,11 @@ class ContactControllerCore extends FrontController
     public function postProcess()
     {
         if (Tools::isSubmit('submitMessage')) {
+            $saveContactKey = $this->context->cookie->formCleContact;
```
Comment (xBorderie) [inline controllers/front/ContactController.php]: Could you remove the French? `formCleContact` is not really international ;)

Comment (hibatallahAouadni) [reply controllers/front/ContactController.php]: thanks @xBorderie ;)

File: controllers/front/ContactController.php [hunk 2]
```diff
@@ -247,9 +251,13 @@ public function initContent()
...
         }
 
+        $contactKey = md5(uniqid(microtime(), true));
+        $this->context->cookie->__set('contactFormKey', $contactKey);
+
         $this->context->smarty->assign(array(
             'contacts' => Contact::getContacts($this->context->language->id),
-            'message' => html_entity_decode(Tools::getValue('message'))
+            'message' => html_entity_decode(Tools::getValue('message')),
+            'contactKey' => $contactKey
```
Comment (LittleBigDev) [inline controllers/front/ContactController.php]: Please add a comma even after the last array item

**Review by LittleBigDev — CHANGES_REQUESTED** [review]

File: controllers/front/ContactController.php (hunk 2 above)
Comment (hibatallahAouadni) [reply controllers/front/ContactController.php]: @LittleBigDev done ;)

**Review by LittleBigDev — APPROVED** [review]

**Comment by marionf** [comment]
Hello @hibatallahAouadni 

I can't send message anymore

capture du 2017-08-28 15-38-34

**Comment by hibatallahAouadni** [comment]
Hello @marionf 
I'm really sorry but I didn't manage to reproduce your issue, I send 3 mails just right now and all have been successfully sent (see screenshots) and I tried it with customers and visitors.
- success message:
contactanos buen gusto
- received mails (BO):
sav buen gusto

**Comment by marionf** [comment]
Hello @hibatallahAouadni 
It's ok, it's working.
Thanks!

**Comment by maximebiloe** [comment]
Thanks @hibatallahAouadni

**Comment by yaya85** [comment]
Verry good solotion, but in the  1.6.1.17 version it seems to give an error on the contact page when you try to send >http://prntscr.com/gumu69

I added the 2 rows to the contact page form, but still having an error.
When i add the ContactController.php from 1.6.1.16 it seems to work.

Do i need to change or add something to make it work in 1.6.1.17 ?

**Comment by bumarius** [comment]
Hello,

please DO NOT forget also to modify contact.tpl
```
**<input type="text" name="url" value="" class="hidden" />**
**<input type="hidden" name="contactKey" value="{$contactKey}" />**
<button type="submit" name="submitMessage" id="submitMessage" class="xxxxxxxxxxxxxxxx" style="margin-top: 0 !important;">
```

best regards

**Comment by yaya85** [comment]
Hi Bumarius,

You mean contact-form.tpl right? I already added this 2 lines of code. > http://prntscr.com/guv1i9

Is there anything else what we have to change?

Thanks!

**Comment by gregzawadzki** [comment]
I developed module for this today using reCaptcha "I'm not a robot" : https://github.com/seigieu/seigisecurecontact
You can get it from git or there is link where you can download redy-to-install zip.

**Comment by ghost** [comment]
Same thing as @marionf mentioned. 
An error occurred while sending the message in FO, but message is available in BO.
Presta 1.6.1.10
It was related with SMTP missconfiguration... Customer change it's password and error occur in BO because confirmation email was unable to send.
________________________________
@yaya85 check email configuration and/or SMTP configuration/password.

@seigieu thanks for module but reCaptcha will not work in this case. Spammers will pass through reCaptcha validation...

**Comment by gregzawadzki** [comment]
@PrestaShark Which "case" you mean? What does not work?

**Review by crsisti1 — COMMENTED** [review]
what changes do i have to make?i don't understand

**Comment by DaoPuig** [comment]
I used another solution and works for me.
 
ContactController.php inside the PostProcess function after the first if (Tools::isSubmit('sumbitMessage'))

// Get contact email 
$em = trim(Tools::getValue('from'));
// We will use it to check russian mails
$length = strlen('.ru');
$message = Tools::getValue('message');

// Check if the mail contact is a russian email
if (substr($em, -$length) === '.ru') {
   $this->errors[] = Tools::displayError('Invalid email address.');
// Check if content is written in russian
} elseif ($this->isRussian($message)) {
     $this->errors[] = Tools::displayError('Invalid content.');
} .....
// Function to check if the message is written in russian
private function isRussian($text) {
                return preg_match('/[А-Яа-яЁё]/u', $text);
        }

**Comment by gregzawadzki** [comment]
@DaoPuig That's putting all russians in one bag. And I had problems with Chinese spam, so if you want to generalize protection, add chinese letters also. And would also add checking for links and domains in message :)

**Comment by DaoPuig** [comment]
@seigieu You are right, my bad. I came from that thread [link](http://forge.prestashop.com/browse/PSCSX-9781?focusedCommentId=163967&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163967) about **Russian messages spamming.**
Actually, I had only spam from Russians and I tried to help with a fast hack :D

**Comment by gregzawadzki** [comment]
@DaoPuig think fastest and best solution is one I posted above. No generalization and no spam. Almost cannot be bypassed, since google updates it on regular basis.

**Comment by ghost** [comment]
Guys. Just for mention. Solution posted in this pull request failed to block russian spammers. Upgraded ContactController and tpl file and mail.ru spam pass through contact form easily even paired with ReCaptcha. WTF

**Comment by alegout** [comment]
Hi @PrestaShark, i'm currently working on a PR to limit these spams issues on the contactform. 
For a robust solution, consider using a captcha solution like ReCaptcha (We can not add one because we can not require users to install FreeType). 
If you are on 1.6, please disable sendtofriends module. 
If the problem's not solved, are on 1.7 and have Recaptcha running, please create an issue and give us more details please, we'll check ASAP. 

Best regards

**Comment by ghost** [comment]
Hi. Details are clear. Changes in this pull request not work anymore. Spamers rebuild their scripts.
1.6.1.14 with pull fixes = failed to stop mail.ru spam

**Comment by alegout** [comment]
You said that you have a ReCaptcha module, can you please test with javascript disabled if your module is working?

Thank you

**Comment by DaoPuig** [comment]
@PrestaShark My solution is still working at least in my shop, as a fast hack you can try it.

**Comment by ghost** [comment]
@DaoPuig yes i trying it too but with Your hack no email are send anymore and notifications in BO are not work. Messages are saved in DB. Any ideas?

`if (Tools::isSubmit('submitMessage')) {
            $saveContactKey = $this->context->cookie->contactFormKey;
            $extension = array('.txt', '.rtf', '.doc', '.docx', '.pdf', '.zip', '.png', '.jpeg', '.gif', '.jpg');
            $file_attachment = Tools::fileAttachment('fileUpload');
			$em = trim(Tools::getValue('from'));
			$length = strlen('.ru');
            $message = Tools::getValue('message'); // Html entities is not usefull, iscleanHtml check there is no bad html tags.
            $url = Tools::getValue('url');
            if (!($from = trim(Tools::getValue('from'))) || !Validate::isEmail($from)) {
                $this->errors[] = Tools::displayError('Invalid email address.');
			} elseif (substr($em, -$length) === '.ru') {
				$this->errors[] = Tools::displayError('Invalid email address.');
			} elseif ($this->isRussian($message)) {
				$this->errors[] = Tools::displayError('Invalid content.');
			} elseif (!$message) {
                $this->errors[] = Tools::displayError('The message cannot be blank.');
            } elseif (!Validate::isCleanHtml($message)) {
                $this->errors[] = Tools::displayError('Invalid message');
            } elseif (!($id_contact = (int)Tools::getValue('id_contact')) || !(Validate::isLoadedObject($contact = new Contact($id_contact, $this->context->language->id)))) {
                $this->errors[] = Tools::displayError('Please select a subject from the list provided. ');
            } elseif (!empty($file_attachment['name']) && $file_attachment['error'] != 0) {
                $this->errors[] = Tools::displayError('An error occurred during the file-upload process.');
            } elseif (!empty($file_attachment['name']) && !in_array(Tools::strtolower(substr($file_attachment['name'], -4)), $extension) && !in_array(Tools::strtolower(substr($file_attachment['name'], -5)), $extension)) {
                $this->errors[] = Tools::displayError('Bad file extension');
            } elseif ($url === false || !empty($url) || $saveContactKey != (Tools::getValue('contactKey'))) {
                $this->errors[] = Tools::displayError('An error occurred while sending the message.');
            } else {`

**Comment by gregzawadzki** [comment]
@alegout If you mean my module... then no. Can see no point in no-javascript support. It's like 0.001% of users. Javascript is everywhere (fix me, if I'm wrong).

**Comment by jchelink** [comment]
@DaoPuig Hi, your solution works for me for now :) Thanks
v1.6.1.1

**Comment by alegout** [comment]
yes you are right, but a bot just call the url, do not even see your recaptcha :)

**Comment by alegout** [comment]
ok @seigieu my bad, i saw that you valid the captcha in back-end, so 👍 i tought that is was only a javascript validation like another module i saw. sorry :)

**Comment by CrimsonLockhart** [comment]
For those who have no idea what they are doing like me, there were a few syntax errors that kept the form from opening on 1.6.1.7.  Here is the copy of DaoPuigs deal that I fixed.  It got the form to open again and I am waiting to see if it gets these mfers to stop for now.  I have tried captcha, and pretty much every other solution on here but this one.  Looking promising.  :)

//Start russian bs
			// Get contact email
$em = trim(Tools::getValue('from'));
// We will use it to check russian mails
$length = strlen('.ru');
$message = Tools::getValue('message');

// Check if the mail contact is a russian email
if (substr($em, -$length) === '.ru') {
$this->errors[] = Tools::displayError('Invalid email address.');
// Check if content is written in russian
} elseif ($this->isRussian($message)) {
$this->errors[] = Tools::displayError('Invalid content.');
}
// Function to check if the message is written in russian
function isRussian($text) {
return preg_match('/[А-Яа-яЁё]/u', $text);
}
			//End russian bs

**Comment by jocel1** [comment]
FYI https://github.com/nenes25/eicaptcha seems to work quite well to prevent spam by using recaptcha (0.4.x branch for prestashop 1.6.x, 2.x for prestashop 1.7.x)

**Comment by mittermichal** [comment]
FYI this commit breaks all older 1.6 custom templates since they miss 
`<input type="text" name="url" value="" class="hidden" />
<input type="hidden" name="contactKey" value="{$contactKey}" />`

**Comment by fire2** [comment]
I updated PS to 1.6.1.18 (from 1.6.1.16) and this change was included. So I updated my theme and added those 2 hidden inputs to the contact form. However I still get an error when trying to send the form because the submitted contactkey and the key in the form do not match. I added a die() with both values in that validation check and it fails, I really don't understand why. 

You can check there: tienda(dot)deflamenco(dot)com/en/contact-us 

What could be happening?

**Comment by tzak902** [comment]
Hi guys, what about 1.7.2.5 branch

**Comment by mhd6** [comment]
Hello my friend, I did not understand or put the secret key ? Thanks you

https://github.com/PrestaShop/PrestaShop/pull/8168/files/1482acec6e91a6e694b9638d42ac69be391dd7fc

**Review by mkinternet — COMMENTED** [review]
This fix is not good enough. Cookie has to be erased every time form submit is confirmed. Spam bot can use created cookie many times.

Add to the 211 line
$this->context->cookie->__unset('contactFormKey');

before:
$this->context->smarty->assign('confirmation', 1);

**Pull request merged**